src/Psl/Type/Internal/F32Type.php

```diff
@@ -39,15 +39,13 @@ public function matches(mixed $value): bool
      */
     public function coerce(mixed $value): float
     {
-        $float = Type\float()
-            ->withTrace($this->getTrace()->withFrame($this->toString()))
-            ->coerce($value);
+        $float = Type\float()->coerce($value);
```

Comment:
I'm gonna wash myself later, but this might be a good solution?
Less objects, so should be faster if we implement this on every type function. (assuming none of them has leaking state / memory)

```php
namespace Psl\Type;

use Psl\Type\Internal\FloatType;

/**
 * @return TypeInterface<float>
 */
function float(): TypeInterface
{
    /** @var FloatType|null $instance */
    static $instance = null;
    if (null === $instance) {
        $instance = new Internal\FloatType();
    }

    return $instance;
}

```

Reply:
Let's discuss memoization elsewhere: I'm fairly sure that we can squeeze quite a lot of performance out of this, in future.

This is also feasible now due to your work removing traces from all type objects, which now effectively become pure :)